Fix docker context calculation re: .dockerignore

ianfixes opened the pull request:
* Fixes #576 

This patch implements the logic described [in the Dockerfile reference for `.dockerignore`](https://docs.docker.com/engine/reference/builder/#dockerignore-file).

> Lines starting with `!` (exclamation mark) can be used to make exceptions to exclusions. The following is an example `.dockerignore` file that uses this mechanism:
> ```
> *.md
> !README.md
> ```
> All markdown files except `README.md` are excluded from the context.
> 
> The placement of ! exception rules influences the behavior: the last line of the `.dockerignore` that matches a particular file determines whether it is included or excluded. Consider the following example:
> 
> ```
> *.md
> !README*.md
> README-secret.md
> ```
> 
> No markdown files are included in the context except `README` files other than`README-secret.md`.

ianfixes commented:
@tlunter could you please approve running workflows on this PR so that I can get CI results?

ianfixes commented:
@nylkotuby could you please approve running workflows on this PR so that I can get CI results?

nylkotuby commented:
@ianfixes I'm no longer a maintainer on this repository, sorry!

ls-todd-lunter commented:
@ianfixes approved. Sorry, we've migrated accounts over here so I missed the tag.

ianfixes commented:
Thanks to both of you!  

Also, small world: Todd I think when I toured Upserve you showed me your Amtrak status tool!

ianfixes commented:
I think I need the approval one more time.  There must be some approval tied to a commit that goes away if I force-push my branch.  Added unit tests and fixed the ABC complexity issue.

ianfixes commented:
All tests are now passing, I look forward to any review comments

ianfixes commented:
@tlunter can I get a second approval here?

ianfixes commented:
What timeline am I looking at for having this reviewed and (hopefully) released?  This is an honest question, not an expression of urgency -- my team is just trying to plan around it.

ls-todd-lunter commented on lib/docker/util.rb:
```diff
@@ -259,21 +284,22 @@ def build_config_header(credentials)
-    return unless files.include?(ignore)
-    ignored_files(directory, ignore).each { |f| files.delete(f) }
+  def glob_all_files(pattern)
+    (Dir.glob(pattern, File::FNM_DOTMATCH) - ['..', '.']).reject { |p| p.end_with?("/.") }
```
Is the change for this function needed? Specifically the reject for lines ending in `/.`. Just wondering when this would happen.

ianfixes replied:
Sure -- just added some debugging lines to find out where this happens.  

For a glob of `/Users/ianfixes/Development/docker-api/spec/docker/../fixtures/top/**/*`, one of the returned values is `/Users/ianfixes/Development/docker-api/spec/docker/../fixtures/top/.`

ls-todd-lunter commented on lib/docker/util.rb:
```diff
@@ -142,10 +144,33 @@ def create_dir_tar(directory)
+
+    # iterate over non-comments and non-blanks from the dockerignore file,
+    # starting with the initial glob as working set
+    File.read(dockerignore).each_line.map(&:strip)
```
Can we move each of these to new lines? Would prefer if we're chaining vertically, to chain it all that way.

ianfixes commented on lib/docker/util.rb:
```diff
@@ -1,8 +1,10 @@
 module Docker::Util
   # http://www.tldp.org/LDP/GNU-Linux-Tools-Summary/html/x11655.htm#STANDARD-WILDCARDS
-  GLOB_WILDCARDS = /[\?\*\[\{]/
+  GLOB_WILDCARDS = /[\?\*\[\{\]\}]/
```
This is no longer used anywhere (I factored out the need for it while reducing ABC complexity), but I wasn't sure whether to delete it

ls-todd-lunter replied:
That makes sense. Let's leave it, but maybe we can add a comment that it's unused now?

ls-todd-lunter commented on lib/docker/util.rb:
```diff
@@ -259,21 +286,23 @@ def build_config_header(credentials)
+  # specifically: matched directories are considered a recursive match
+  def dockerignore_compatible_glob(pattern)
+    begin
+      some_dirs, some_files = glob_all_files(pattern).partition { |f| File.directory?(f) }
```
I tried to see if we can use `File.fnmatch?` instead of needing to re-glob each time, but apparently the rules for that don't work. `x/**/y` doesn't match `x/y` while it does for `Dir.glob`... That's regardless of the flags passed to `fnmatch?` as well.

ianfixes replied:
Does that mean some action is necessary here?  Am I missing a test case?

ls-todd-lunter replied:
Sorry, nope, was just sharing another avenue that doesn't seem possible right now.

ls-todd-lunter commented:
Released in v2.2.0.

ianfixes commented:
Awesome, thank you!
